File: Dynamic_Programming/min_path_sum.js
// https://leetcode.com/problems/triangle/submissions/

// Runtime: 68 ms, faster than 99.11 % of JavaScript online submissions for Triangle.
// Memory Usage: 39.3 MB, less than 6.83 % of JavaScript online submissions for Triangle.

var minimumTotal = function (triangle, memo = triangle.map(layer => layer.slice().fill(undefined)), i = 0, j = 0) {

    if (i >= triangle.length) return 0; if (memo[i][j]) return memo[i][j];

    const right = Math.min(j + 1, triangle[i].length - 1)

    return memo[i][j] = Math.min(
        triangle[i][j] + minimumTotal(triangle, memo, i + 1, j),
        triangle[i][right] + minimumTotal(triangle, memo, i + 1, right)
    )

};

File: Dynamic_Programming/min_path_sum.rb
# https://leetcode.com/problems/triangle/

# Runtime: 44 ms, faster than 83.33% of Ruby online submissions for Triangle.
# Memory Usage: 210.1 MB, less than 16.67% of Ruby online submissions for Triangle.

def minimum_total(triangle)
    for i in (triangle.length - 2).downto(0)
        for j in (0...triangle[i].length)
            triangle[i][j] = triangle[i][j] + [triangle[i + 1][j], triangle[i + 1][j + 1]].min
        end
    end
    
    return triangle[0][0]
end